src/reducer.ts

```diff
@@ -0,0 +1,46 @@
+export interface RequestState<Data> {
+  data: Data | null;
+  loading: boolean;
+  error: Error | null;
+}
+
+type Action<Data> =
+  | { type: 'REQUEST_INIT' }
+  | { type: 'REQUEST_SUCCESS'; payload: Data }
+  | { type: 'REQUEST_FAILED'; payload: Error };
+
+export const initialState = {
+  data: null,
+  error: null,
+  loading: false,
+};
+
+export default <Data>() => (
+  state: RequestState<Data>,
+  action: Action<Data>
+): RequestState<Data> => {
+  switch (action.type) {
+    case 'REQUEST_INIT':
+      return {
+        ...state,
+        data: null,
```

Comment:
Should `data` be set to `null` on `REQUEST_INIT`?

I was thinking about the use case where `data` has been loaded already, and there's some kind of "Refresh" button the user can click to fetch latest data - I'm not sure we would want to naively clear out the `data` since consumers may want to hold onto it while new data is being requested.

If they don't want to display the current `data` while `loading`, they can do something like this:

```js
if (data && !loading) {
  return <div>{data}</div>
}
```

So holding onto the `data` would account for both of those use cases.